include/ghex/structured/cubed_sphere/field_descriptor.hpp

```diff
@@ -0,0 +1,504 @@
+                    static constexpr std::size_t block_dim = 128;
+                    // number of values to pack/unpack per thread
+                    static constexpr std::size_t elements_per_thread = 1;
+
+                    template<typename PackIterationSpace>
+                    static void pack(PackIterationSpace&& pack_is, void* arg) {
+                        using coordinate_type = typename PackIterationSpace::coordinate_t;
+                        auto stream_ptr = reinterpret_cast<cudaStream_t*>(arg);
+                        const std::size_t num_threads = (pack_is.m_buffer_desc.m_size + elements_per_thread-1)
+                            /elements_per_thread;
+                        const std::size_t num_blocks = (num_threads+block_dim-1)/block_dim;
+                        pack_kernel<LMap><<<num_blocks,block_dim,0,*stream_ptr>>>(pack_is, elements_per_thread);
+                    }
+
+                    template<typename UnPackIterationSpace>
+                    static void unpack(UnPackIterationSpace&& unpack_is, void* arg) {
+                        using coordinate_type = typename UnPackIterationSpace::coordinate_t;
+                        auto stream_ptr = reinterpret_cast<cudaStream_t*>(arg);
+                        const std::size_t num_threads = (unpack_is.m_buffer_desc.m_size + elements_per_thread-1)
+                            /elements_per_thread;
+                        const std::size_t num_blocks = (num_threads+block_dim-1)/block_dim;
+                        unpack_kernel<LMap><<<num_blocks,block_dim,0,*stream_ptr>>>(unpack_is, elements_per_thread);
+                    }
+                };
+#endif
+
+                /** @brief Describes a field spanning a subdomain of a cubed-sphere tile
+                  * @tparam T value type
+                  * @tparam Arch Architecture type
+                  * @tparam ComponentOrder Data layout order parameter (3 == stride 1, 0 == largest stride)
+                  * @tparam XOrder Data layout order parameter (3 == stride 1, 0 == largest stride)
+                  * @tparam YOrder Data layout order parameter (3 == stride 1, 0 == largest stride)
+                  * @tparam ZOrder Data layout order parameter (3 == stride 1, 0 == largest stride)*/
+                template<typename T, typename Arch, int ComponentOrder=0, int XOrder=3, int YOrder=2, int ZOrder=1>
+                class field_descriptor {
+                public: // member types
+                    using value_type             = T;
+                    using arch_type              = Arch;
+                    using device_id_type         = typename arch_traits<arch_type>::device_id_type;
+                    using domain_descriptor_type = domain_descriptor;
+                    using dimension              = domain_descriptor_type::dimension;
+                    using layout_map             = ::gridtools::layout_map<ComponentOrder,XOrder,YOrder,ZOrder>;
```

Comment:
Here we assume component order if the first component of the array accessor. Is this a specific FV3 convention?

Reply:
no this is not a specific FV3 convention - I chose it (also it is now the last in the list). However, it is used internally only anyways: the user would rarely use our field_descriptor directly.